### PR TITLE
[2854] extension property support

### PR DESCRIPTION
**Pull request by alainosorio**

### CODE REVIEW

**Comment by yanavasileva** [comment]
Hi @alainosorio,

Thank you for your contribution.
I created camunda/camunda-bpm-platform/issues/2854 to track the progress of this request.
I will send the ticket for a decision, please keep in mind that it could take a while to get back to you because of other team responsibilities. Thank you for your patience in advance. 
In case you would like to speed up the process you can already think about adding tests and documentation in the contribution.

Best regards,
Yana

**Comment by yanavasileva** [comment]
@alainosorio 
To extend the documentation, please add the new property to bottom of client.subscribeTo table:
https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#clientsubscribetopic-options-handler
The existing test cases can be found here: https://github.com/camunda/camunda-external-task-client-js/blob/master/lib/Client.test.js#L209-L242

In case of any questions, please let us know.

**Comment by alainosorio** [comment]
thanks @yanavasileva, will do and let you know

File: lib/Client.test.js [hunk 1]
```diff
@@ -228,7 +228,8 @@ describe("Client", () => {
         processDefinitionVersionTag: "versionTag",
         tenantIdIn: ["tenantId"],
         withoutTenantId: true,
-        localVariables: true
+        localVariables: true,
+        extensionProperties: true,
```
Comment (yanavasileva) [inline lib/Client.test.js:232]: Redundant comma:
```suggestion
        extensionProperties: true
```

**Review by yanavasileva — COMMENTED** [review]
@alainosorio , you can add test data here:
https://github.com/camunda/camunda-external-task-client-js/blob/master/lib/__snapshots__/Client.test.js.snap#L25-L58

File: docs/Client.md [hunk 2]
```diff
@@ -76,7 +76,7 @@ The currently supported options are:
 | tenantIdIn | A value which allows to filter tasks based on tenant ids         | string |          |                                                       |
 | withoutTenantId | A value which allows to filter tasks without tenant id                              | boolean |         |                                                       |
 | localVariables | A value which allow to fetch only local variables     | boolean |          |                                                       |
-
+| extensionProperties | A value which allow to fetch extensionProperties     | boolean |          |                                                       |
```
Comment (tasso94) [inline docs/Client.md:79]:
```suggestion
| includeExtensionProperties | A value which allow to fetch `extensionProperties`     | boolean |          |                                                       |
```

File: lib/Client.js [hunk 3]
```diff
@@ -196,7 +196,8 @@ class Client extends events {
           processVariables,
           tenantIdIn,
           withoutTenantId,
-          localVariables
+          localVariables,
+          extensionProperties
```
Comment (tasso94) [inline lib/Client.js:200]:
```suggestion
          includeExtensionProperties
```

File: lib/Client.js [hunk 4]
```diff
@@ -245,6 +246,10 @@ class Client extends events {
           topic.localVariables = localVariables;
         }
 
+        if (!isUndefinedOrNull(extensionProperties)) {
+          topic.extensionProperties = extensionProperties;
```
Comment (tasso94) [inline lib/Client.js:250]:
```suggestion
        if (!isUndefinedOrNull(includeExtensionProperties)) {
          topic.includeExtensionProperties = includeExtensionProperties;
```

File: lib/Client.test.js [hunk 5]
```diff
@@ -228,7 +228,8 @@ describe("Client", () => {
         processDefinitionVersionTag: "versionTag",
         tenantIdIn: ["tenantId"],
         withoutTenantId: true,
-        localVariables: true
+        localVariables: true,
+        extensionProperties: ["property1", "property2"]
```
Comment (tasso94) [inline lib/Client.test.js:232]:
```suggestion        
        includeExtensionProperties: true
```

File: lib/Client.test.js [hunk 6]
```diff
@@ -374,6 +375,19 @@ describe("Client", () => {
       );
     });
 
+    test("should subscribe to topic based on extensionProperties ", () => {
```
Comment (tasso94) [inline lib/Client.test.js:378]:
```suggestion
    test("should subscribe to topic with includeExtensionProperties = true", () => {
```

File: lib/Client.test.js [hunk 7]
```diff
@@ -374,6 +375,19 @@ describe("Client", () => {
...
+      // given
+      const footopicSubscription = client.subscribe(
+        "foo",
+        customConfig,
+        fooWork
+      );
+      // then
+      expect(footopicSubscription.extensionProperties).toBe(
+        customConfig.extensionProperties
```
Comment (tasso94) [inline lib/Client.test.js:387]:
```suggestion
      expect(footopicSubscription.includeExtensionProperties).toBe(
        customConfig.includeExtensionProperties
```

File: lib/__snapshots__/Client.test.js.snap [hunk 8]
```diff
@@ -28,6 +28,10 @@ Array [
     "maxTasks": 3,
     "topics": Array [
       Object {
+        "extensionProperties": Array [
+          "property1",
+          "property2",
+        ],
```
Comment (tasso94) [inline lib/__snapshots__/Client.test.js.snap:34]:
```suggestion
        "includeExtensionProperties": true,
```

**Review by tasso94 — CHANGES_REQUESTED** [review]
* The subscription option to fetch extension properties is called `includeExtensionProperties`.
* It is not possible to filter for extension properties – only no or all extension properties will be fetched.

**Comment by yanavasileva** [comment]
@alainosorio  thank you for implementing the review hints. I will have a look and let you know once I merge the contribution as I believe we are on a good track with the PR already.

**Comment by alainosorio** [comment]
> @alainosorio thank you for implementing the review hints. I will have a look and let you know once I merge the contribution as I believe we are on a good track with the PR already.

Thanks!

**Review by yanavasileva — APPROVED** [review]

**Comment by yanavasileva** [comment]
Hi @alainosorio,

I adjusted the commit message according to our [CONTRIBUTING](https://github.com/camunda/camunda-bpm-platform/blob/master/CONTRIBUTING.md#commit-message-conventions) guidelines and merged the code with 170192f.
Thank you and I see you in your next contribution!

Best regards,
Yana

**Comment by alainosorio** [comment]
> Hi @alainosorio,
> 
> I adjusted the commit message according to our [CONTRIBUTING](https://github.com/camunda/camunda-bpm-platform/blob/master/CONTRIBUTING.md#commit-message-conventions) guidelines and merged the code with [170192f](https://github.com/camunda/camunda-external-task-client-js/commit/170192fd1b169229cc860e638b31f0216b3ba3a4). Thank you and I see you in your next contribution!
> 
> Best regards, Yana

Thanks!

**Pull request closed without merging**